controllers/slice/namespaces.go

```diff
@@ -205,67 +202,182 @@ func (r *SliceReconciler) reconcileAllowedNamespaces(ctx context.Context, slice
+	// Sweep the existing namespaces again to unbind any namespace that was not found in the configured list
+	for existingAllowedNs, _ := range existingAllowedNsMap {
+		if !existingAllowedNsMap[existingAllowedNs].marked {
+			err := r.unbindAllowedNamespace(ctx, existingAllowedNs, slice.Name)
+			if err != nil {
+				log.Error(err, "Failed to unbind namespace from slice", "namespace", existingAllowedNs)
+				return err
+			}
+			log.Info("unbind allowed namespace success", "namespace", existingAllowedNs)
+			statusChanged = true
+		}
+	}
+	if statusChanged {
+		slice.Status.AllowedNamespaces = labeledAllowedNsList
+		err := r.Status().Update(ctx, slice)
+		if err != nil {
+			log.Error(err, "Failed to update slice status")
+			return err
+		}
 	}
 	return nil
 }
 
+func (r *SliceReconciler) annotateNamespace(ctx context.Context, slice *kubeslicev1beta1.Slice, allowedNamespace *corev1.Namespace) (bool, error) {
+	annotations := allowedNamespace.GetAnnotations()
+	if annotations == nil {
+		//add first annotation
+		annotations = map[string]string{}
+		annotations[AllowedNamespaceAnnotationKey] = slice.Name
+		allowedNamespace.ObjectMeta.Annotations = annotations
+		return true, r.Update(ctx, allowedNamespace)
+	} else if _, ok := annotations[AllowedNamespaceAnnotationKey]; !ok {
+		//annotations not nil, but AllowedNamespaceAnnotationKey not present
+		annotations[AllowedNamespaceAnnotationKey] = slice.Name
+		allowedNamespace.ObjectMeta.Annotations = annotations
+		return true, r.Update(ctx, allowedNamespace)
+	}
+	// AllowedNamespaceAnnotationKey present, append the comma seperated sliceName to value
+	// eg : kubeslice.io/trafficAllowedToSlices: "slice-1,slice-2,slice-3"
+	v := annotations[AllowedNamespaceAnnotationKey]
+	// create an array of slices and check if the slice name already exists
+	a := strings.Split(v, ",")
+	if !exists(a, slice.Name) {
+		annotations[AllowedNamespaceAnnotationKey] = v + "," + slice.Name
+		allowedNamespace.ObjectMeta.Annotations = annotations
+		return true, r.Update(ctx, allowedNamespace)
+	}
+	return false, nil
+}
+
+// unbindAllowedNamespace will remove the slice name from annotation
+// in case it the last slice, it will also remove kubeslice Allowed Namespace label
+func (r *SliceReconciler) unbindAllowedNamespace(ctx context.Context, allowedNs, sliceName string) error {
+	log := logger.FromContext(ctx).WithValues("type", "allowedNamespace")
+	namespace := &corev1.Namespace{}
+	err := r.Get(ctx, types.NamespacedName{Name: allowedNs}, namespace)
+	if err != nil {
+		log.Error(err, "NS unbind: Failed to find namespace", "namespace", allowedNs)
+		return err
+	}
+
+	//early exit if annotations are nil or AllowedNamespaceAnnotationKey is not present
+	annotations := namespace.GetAnnotations()
+	if annotations == nil {
+		return nil
+	}
+	v := annotations[AllowedNamespaceAnnotationKey]
+	if v == "" {
+		return nil
+	}
+
+	a := strings.Split(v, ",")
+	if len(a) == 1 && exists(a, sliceName) {
```

Comment:
Hey @bharath-avesha , looks like a small problem?
eg: kubeslice.io/namespace: slice1,slice2,slice3
if the namespace is offboarded from slice1
toDeleteSlice = 0
in this case according to this block , it would be deleting all the annotations

Reply:
a := strings.Split(v, ",")
if exists() {
if(len(a)) == 1 {
//remove annotation and label
}else{
//remove annotation
}
}

how does this looks? basically it would be in a single block

Reply:
Yes, i was wrong, it needs a len check, not the index. Lets do that.